app/controllers/application_controller.rb

```diff
@@ -5,6 +5,126 @@ class ApplicationController < Sinatra::Base
   configure do
     set :public_folder, 'public'
     set :views, 'app/views'
+    enable :sessions
+    set :session_secret, "secret"
+  end
+
+  get '/' do
+    erb :index
+  end
+
+  get '/signup' do
+    if session[:id]
+      redirect to '/tweets'
+    else
+      erb :"users/create_user"
+    end
+  end
+
+  post '/signup' do
+    user = User.new(params)
+    if user.save
+      session[:id] = user.id
+      redirect to '/tweets'
+    else
+      redirect to '/signup'
+    end
+  end
+
+  get '/login' do
+    if session[:id]
+      redirect to '/tweets'
+    else
+      erb :"users/login"
+    end
+  end
+
+  post '/login' do
+    user = User.find_by(username: params[:username])
+    if user && user.authenticate(params[:password])
+      session[:id] = user.id
+      redirect to 'tweets'
+    else
+      redirect to 'login'
+    end
+  end
+
+  get '/logout' do
+    session.clear
+    redirect to '/login'
+  end
+
+  get '/users/:slug' do
+    @user = user.find_by_slug(params[:slug])
+    erb :'users/show'
+  end
+
+  get '/tweets' do
+    if session[:id]
+      @user = User.find(session[:id])
+      @tweets = Tweet.all
+      erb :"tweets/tweets"
+    else
+      redirect to '/login'
+    end
+  end
+
+  get '/tweets/new' do
+    if session[:id]
+      erb :"tweets/create_tweet"
+    else
+      redirect to '/login'
+    end
+  end
+
+  post '/tweets' do
+    tweet = Tweet.create(content: params[:content], user: User.find(session[:id]))
+    if tweet.save
+      redirect to '/tweets'
+    else
+      redirect to '/tweets/new'
+    end
+  end
+
+  get '/tweets/:id' do
+    if session[:id]
+      @tweet = Tweet.find(params[:id])
+      erb :"tweets/show_tweet"
+    else
+      redirect to '/login'
+    end
+  end
+
+  get '/tweets/:id/edit' do
+    if session[:id]
+      @tweet = Tweet.find(params[:id])
+      if session[:id] == @tweet.user.id
+        erb :"tweets/edit_tweet"
+      else
+        redirect to "tweets/#{params[:id]}"
+      end
+    else
+      redirect to '/login'
+    end
+  end
+
+  post '/tweets/:id' do
+    tweet = Tweet.find(params[:id])
+    tweet.content = params[:content]
+    if tweet.save
+      redirect to "/tweets/#{params[:id]}"
+    else
+      redirect to "/tweets/#{params[:id]}/edit"
+    end
+  end
+
+  post '/tweets/:id/delete' do
+    tweet = Tweet.find(params[:id])
+    if session[:id] == tweet.user.id
```

Comment:
Prefer to find tweets that are already only associated to the `current_user` than to manually check the `user_id` foreign key of the tweet against the `current_user.id`

```ruby
tweet = current_user.tweets.find_by(:id => params[:id])
if tweet && tweet.destroy
  redirect "/tweets"
else
  redirect "/tweets/#{tweet.id}"
end
```

We're using the association collection method provided by [`has_many`](http://api.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html#method-i-has_many) to query the users tweets. 

Note the usage of [`find_by`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find-by) vs [`find`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find). Had we used `find` here, ActiveRecord would raise an error in the event that we could not find a tweet associated with this user. `find_by` will allow this action to fail gracefully simply returning nil if the user is trying to delete a tweet that they do not have access to.